src/datasets/packaged_modules/sql/sql.py

```diff
@@ -0,0 +1,97 @@
+from dataclasses import dataclass
+from typing import TYPE_CHECKING, Dict, List, Optional, Tuple, Union
+
+import pandas as pd
+import pyarrow as pa
+
+import datasets
+import datasets.config
+from datasets.features.features import require_storage_cast
+from datasets.table import table_cast
+
+
+if TYPE_CHECKING:
+    import sqlalchemy
+
+
+@dataclass
+class SqlConfig(datasets.BuilderConfig):
+    """BuilderConfig for SQL."""
+
+    sql: Union[str, "sqlalchemy.sql.Selectable"] = None
+    con: str = None
+    index_col: Optional[Union[str, List[str]]] = None
+    coerce_float: bool = True
+    params: Optional[Union[List, Tuple, Dict]] = None
+    parse_dates: Optional[Union[List, Dict]] = None
+    columns: Optional[List[str]] = None
+    chunksize: Optional[int] = 10_000
+    features: Optional[datasets.Features] = None
+
+    def __post_init__(self):
+        assert self.sql is not None, "sql must be specified"
+        assert self.con is not None, "con must be specified"
+
+        if not isinstance(self.con, str):
+            raise ValueError(f"con must be a database URI string, but got {self.con} with type {type(self.con)}.")
+
+    def create_config_id(
+        self,
+        config_kwargs: dict,
+        custom_features: Optional[datasets.Features] = None,
+    ) -> str:
+        # We need to stringify the Selectable object to make its hash deterministic
+
+        # The process of stringifying is explained here: http://docs.sqlalchemy.org/en/latest/faq/sqlexpressions.html
+        sql = config_kwargs["sql"]
+        if not isinstance(sql, str):
+            if not datasets.config.SQLALCHEMY_AVAILABLE:
+                raise ImportError("Please pip install sqlalchemy.")
+
+            import sqlalchemy
+
+            config_kwargs = config_kwargs.copy()
+            engine = sqlalchemy.create_engine(config_kwargs["con"].split("://")[0] + "://")
+            sql_str = str(sql.compile(dialect=engine.dialect))
+            config_kwargs["sql"] = sql_str
```

Comment:
With something like this you can even raise a TypeError if the object is neither a string or a Selectable
```suggestion
            if datasets.config.SQLALCHEMY_AVAILABLE and "sqlalchemy" in sys.modules::
                import sqlalchemy
                
                if isinstance(sql, sqlalchemy.sql.Selectable):
                    config_kwargs = config_kwargs.copy()
                    engine = sqlalchemy.create_engine(config_kwargs["con"].split("://")[0] + "://")
                    sql_str = str(sql.compile(dialect=engine.dialect))
                    config_kwargs["sql"] = sql_str
                else:
                    raise TypeError(f"Supported types for 'sql' are string and sqlalchemy.sql.Selectable but got {type(sql)}: {sql}")
            else:
                raise TypeError(f"Supported types for 'sql' are string and sqlalchemy.sql.Selectable but got {type(sql)}: {sql}")
```